### ISSUE TITLE
Unintuitive dangerous behaviour in VaultLib.Encrypt

### ISSUE THREAD

**Issue by hs-amartin** [ansible/ansible#72703]
---
name: 🐛 Bug report
about: Create a report to help us improve
---

##### SUMMARY
https://github.com/ansible/ansible/blob/35022e13a839d5f59a1c4a254aca12afb124373a/lib/ansible/parsing/vault/__init__.py#L621
My expectation as a developer would be that the vault_id that I pass in to this method would be used by the matcher to filter which is the best secret to use, however that is not the case.  I think this is a very dangerous behaviour as if the ansible.cfg has a config like dev/devpassword.txt,prod/prodpassword.txt, and I call encrypt('value', vault_id=prod), the value will be encrypted with the dev password, but the output object will still label it as part of the prod vault.  

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
VaultLib

##### ANSIBLE VERSION
Note this affects other versions
```
ansible 2.10.3
  config file = /Users/adam/repos/config-management/ansible/ansible.cfg
  configured module search path = ['/Users/adam/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.9/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.9.0 (default, Oct 27 2020, 14:15:17) [Clang 12.0.0 (clang-1200.0.32.21)]
```

##### CONFIGURATION
Redacted to the relevant portion
```paste below
DEFAULT_VAULT_IDENTITY_LIST(/somepath/ansible.cfg) = ['dev@vault/get-pass-dev.sh', 'stage@vault/get-pass-stage.sh', 'prod@vault/get-pass-prod.sh']
```

##### OS / ENVIRONMENT
All

##### STEPS TO REPRODUCE
Run code based on the below. Have not attempted to reproduce using ansible-vault cli.
```python
    loader = DataLoader()
    vault_secret = CLI.setup_vault_secrets(
        loader=loader,
        vault_ids=C.DEFAULT_VAULT_IDENTITY_LIST
    )
    vault = VaultLib(vault_secret)
    vault.encrypt(value, vault_id='prod')
```

##### EXPECTED RESULTS
'value' is encrypted with prod password.

##### ACTUAL RESULTS
'value' is encrypted with dev password.

**Comment by ansibot** [comment]
Files identified in the description:
None

If these files are incorrect, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@hs-amartin: Greetings! Thanks for taking the time to open this issue. In order for the community to handle your issue effectively, we need a bit more information.

Here are the items we could not find in your description:
- issue type
- component name
- ansible version

Please set the description of this issue with an appropriate template from:
https://github.com/ansible/ansible/tree/devel/.github/ISSUE_TEMPLATE

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by RAdamMartin** [comment]
bot_status

**Comment by KatsuragiCSL** [comment]
I cannot reproduce using ansible-vault cli.

**CONFIGURATION**

`DEFAULT_VAULT_IDENTITY_LIST = ['dev@<path>/dev', 'prod@<path>/prod']`

**STEPS TO REPRODUCE**

`ansible-vault encrypt --vault-id prod secrets`
`ansible-vault decrypt --ask-vault-password secrets`
providing prod password successfully decrypted `secrets`

**Comment by RAdamMartin** [comment]
> I cannot reproduce using ansible-vault cli.
> 
> **CONFIGURATION**
> 
> `DEFAULT_VAULT_IDENTITY_LIST = ['dev@<path>/dev', 'prod@<path>/prod']`
> 
> **STEPS TO REPRODUCE**
> 
> `ansible-vault encrypt --vault-id prod secrets`
> `ansible-vault decrypt --ask-vault-password secrets`
> providing prod password successfully decrypted `secrets`

That does appear to work correctly, however it does not appear to care about the configuration at all, it's just taking the direct path (i.e. I think it works for you because you're running the encrypt at the same location as your 'prod' password file).

**Comment by samdoran** [comment]
@hs-amartin In the issue report, you indicated you have not duplicated this using the `ansible-vault` CLI. Can you see test if it works correctly using the CLI? We do not support direct use of the Python API, which is why you may be running into trouble.

needs_info

**Comment by hs-amartin** [comment]
@samdoran I do not believe this is reproducible with ansible-vault since `ansible-vault encrypt --vault-id {id}...` does not inspect the current configuration, and `ansible-vault encrypt --encrypt-vault-id {id}...` appears to behave correctly. However I still think this is worth raising, even as a rejected issue as there are multiple snippets of code out there of people attempting to use this method to programatically manage encrypted strings values given the lack of that functionality inside `ansible-vault`.
